### ISSUE TITLE
SigSegV when refreshing iterator using rocksdbjni

### ISSUE THREAD

**Issue by vmv890** [facebook/rocksdb#12095]
Am I doing something wrong with the `refresh()`? In any case, it probably shouldn't produce a segmentation fault. I should be able to insert data and flush the DB without causing iterator(s) to fail. I apologize if this is expected behavior and it's on me to mitigate it somehow.

### Expected behavior
No Segmentation Fault

### Actual behavior
Segmentation Fault

### Steps to reproduce the behavior
Requires org.junit and rocksdbjni (tested with 8.8.1 and 8.5.3). Tested on jdk17
```java
package rocks;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;
import org.rocksdb.*;

public class RocksIterTest {

    static {
        RocksDB.loadLibrary();
    }

    @Rule
    public TemporaryFolder tempDir = new TemporaryFolder();

    @Test
    public void validateIterRefresh() {
        var dbPath = tempDir.getRoot().getAbsolutePath();

        try (var opts = new Options().setCreateIfMissing(true).setCreateMissingColumnFamilies(true);
            var flushOptions = new FlushOptions();
            var db = RocksDB.open(opts, dbPath))
        {
            for (int i = 0; i < 10; i++)
                db.put(("key." + i).getBytes(), ("value." + i).getBytes());

            db.flush(flushOptions);

            try (var ro = new ReadOptions(); var iter = db.newIterator(ro)) {
                for (iter.seekToFirst(); iter.isValid(); iter.next()) {
                    // Read match
                    String key = new String(iter.key());
                    String val = new String(iter.value());
                    System.out.println(key + " --> " + val);

                    // Insert more data and refresh iterator
                    if (key.equals("key.5")) {
                        for (int i = 10; i < 20; i++)
                            db.put(("key." + i).getBytes(), ("value." + i).getBytes());

                        db.flush(flushOptions);
                        iter.refresh(); // <-- SigSegV after this: rocksdb:MergingIterator::Next()
                    }
                }
            }
        } catch (RocksDBException dbEx) {
            dbEx.printStackTrace();
        }
    }
}
```

**Comment by adamretter** [comment]
> Am I doing something wrong with the refresh()? In any case, it probably shouldn't produce a segmentation fault

The most common cause of a segmentation fault in RocksJava is due to use-after-free. The Java API is a wrapper around the C++ API, so you have to be careful.

> `for (var ro = new ReadOptions(); var iter = db.newIterator(ro)) {`

This code looks invalid to me, is this actually correct?

We need you to share the details of the segmentation fault please before we can tell you much more...

**Comment by vmv890** [comment]
> > Am I doing something wrong with the refresh()? In any case, it probably shouldn't produce a segmentation fault
> 
> The most common cause of a segmentation fault in RocksJava is due to use-after-free. The Java API is a wrapper around the C++ API, so you have to be careful.
> 
> > `for (var ro = new ReadOptions(); var iter = db.newIterator(ro)) {`
> 
> This code looks invalid to me, is this actually correct?

Sorry, I had to re-type the test. It's supposed to say "try" not "for". I fixed the report.
 
> We need you to share the details of the segmentation fault please before we can tell you much more...

Ok, will do shortly.

**Comment by vmv890** [comment]
# JRE version: OpenJDK Runtime Environment Temurin-17.0.9+9 (17.0.9+9) (build 17.0.9+9)
# Java VM: OpenJDK 64-Bit Server VM Temurin-17.0.9+9 (17.0.9+9, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-aarch64)
# Problematic frame:
# C  [librocksdbjni17716741367867510728.jnilib+0x3a6e34]  rocksdb::MergeIteratorBuilder::Finish(rocksdb::ArenaWrappedDBIter*)+0xfec
#
[hs_err_pid36758.log](https://github.com/facebook/rocksdb/files/13465125/hs_err_pid36758.log)

**Comment by adamretter** [comment]
So the relevant parts from your log:

```
#  SIGSEGV (0xb) at pc=0x000000012eb9ee34, pid=36758, tid=8707

...

Stack: [0x000000016d3b8000,0x000000016d5bb000],  sp=0x000000016d5b96d0,  free space=2053k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [librocksdbjni17716741367867510728.jnilib+0x3a6e34]  rocksdb::MergeIteratorBuilder::Finish(rocksdb::ArenaWrappedDBIter*)+0xfec
C  [librocksdbjni17716741367867510728.jnilib+0x3a6ffc]  rocksdb::MergeIteratorBuilder::Finish(rocksdb::ArenaWrappedDBIter*)+0x11b4
C  [librocksdbjni17716741367867510728.jnilib+0x17eb2c]  rocksdb::DBIter::Next()+0x184
j  org.rocksdb.RocksIterator.next0(J)V+0
j  org.rocksdb.AbstractRocksIterator.next()V+26
j  rocksTest.RocksIterTest.validateIterRefresh()V+294
...
```

**Comment by adamretter** [comment]
@vmv890 Is the crash caused by `RocksIterator#next()` caused immediately by the loop after the call to `RocksIterator#refresh()`, or do more iterations of the loop occur after `RocksIterator#refresh()` and before the crash?

**Comment by vmv890** [comment]
> @vmv890 Is the crash caused by `RocksIterator#next()` caused immediately by the loop after the call to `RocksIterator#refresh()`, or do more iterations of the loop occur after `RocksIterator#refresh()` and before the crash?

I do not see any more iterations happen after the refresh.

Here is the debug output (I added a println("Refresh Completed") _after_ refresh call). Everything stops after key.5. The iterator becomes "damaged".

```
key.0 --> value.0
key.1 --> value.1
key.2 --> value.2
key.3 --> value.3
key.4 --> value.4
key.5 --> value.5
Refresh Completed
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000013539ee34, pid=39465, tid=8963
```

**Comment by ajkr** [comment]
`Refresh()` invalidates an iterator: https://github.com/facebook/rocksdb/blob/f6fd4b9dbd15dba36f7e5ad23de407b5c26b1460/include/rocksdb/iterator.h#L111

You will need to call a `Seek*()` function to get the iterator back into a valid state before calling a function that assumes the state is already valid, like `Next()`.

**Comment by vmv890** [comment]
> `Refresh()` invalidates an iterator:
> 
> https://github.com/facebook/rocksdb/blob/f6fd4b9dbd15dba36f7e5ad23de407b5c26b1460/include/rocksdb/iterator.h#L111
> 
> You will need to call a `Seek*()` function to get the iterator back into a valid state before calling a function that assumes the state is already valid, like `Next()`.

I see. Thank you.

**Comment by adamretter** [comment]
@vmv890 I also see that it the doc is present in the RocksJava API - https://github.com/facebook/rocksdb/blob/f6fd4b9dbd15dba36f7e5ad23de407b5c26b1460/java/src/main/java/org/rocksdb/RocksIteratorInterface.java#L119

@ajkr I think it could be clearer though, I will send a PR to update both...

**Comment by vmv890** [comment]
> @vmv890 I also see that it the doc is present in the RocksJava API -
> 
> https://github.com/facebook/rocksdb/blob/f6fd4b9dbd15dba36f7e5ad23de407b5c26b1460/java/src/main/java/org/rocksdb/RocksIteratorInterface.java#L119
> 
> @ajkr I think it could be clearer though, I will send a PR to update both...

Thank you. I completely missed that comment, because when I clicked to the parent, it sent me to the AbstractRocksIterator (which did not have comments. Not an excuse, just a reason :), and I did not click through to the interface (I should have). I checked the wiki https://github.com/facebook/rocksdb/wiki/Iterator and that did not imply any kind of invalidation.

Also, if you change the code to comment out the flush:

```
// db.flush(flushOptions);
iter.refresh();
```

during the update in the test I posted, the test completes and no SigSegV is generated. If the iterator is supposed to be invalidated, I shouldn't be able to do that. It seems that it iterator is more in an "inconsistent" state. Sometimes invalidated, sometimes not, which is why I thought there was something wrong.

Shouldn't this method rather be called reset() or invalidate() ? Refresh doesn't seem to imply that it would be invalidated. Not a big deal, I think I can easily work around this. I realize invalidate() was used for other things, like destruction of objects, so I realize it's not as simple as that.

Again, thank you for your help.
